sklearn/tree/tests/test_tree.py

```diff
@@ -2225,3 +2227,168 @@ def get_joblib_pickle_non_native_endianness():
+
+    clf = DecisionTreeClassifier(random_state=0, max_depth=3)
+    clf.fit(X, y)
+    score = clf.score(X, y)
+
+    def get_different_bitness_joblib_pickle():
+        f = io.BytesIO()
+        p = NumpyPickler(f)
+        p.dispatch_table = copyreg.dispatch_table.copy()
+        p.dispatch_table[CythonTree] = reduce_different_bitness_tree
+
+        p.dump(clf)
+        f.seek(0)
+        return f
+
+    new_clf = joblib.load(get_different_bitness_joblib_pickle())
+    new_score = new_clf.score(X, y)
+    assert np.isclose(score, new_score)
+
+
+def test_check_n_classes():
+    expected_dtype = np.dtype(np.int32) if _IS_32BIT else np.dtype(np.int64)
+    allowed_dtypes = [np.dtype(np.int32), np.dtype(np.int64)]
+    allowed_dtypes += [dt.newbyteorder() for dt in allowed_dtypes]
+
+    n_classes = np.array([0, 1], dtype=expected_dtype)
+    for dt in allowed_dtypes:
+        check_n_classes(n_classes.astype(dt), expected_dtype)
+
+    with pytest.raises(ValueError, match="Wrong dimensions.+n_classes"):
+        wrong_dim_n_classes = np.array([[0, 1]], dtype=expected_dtype)
+        check_n_classes(wrong_dim_n_classes, expected_dtype)
+
+    with pytest.raises(ValueError, match="n_classes.+incompatible dtype"):
+        wrong_dtype_n_classes = n_classes.astype(np.float64)
+        check_n_classes(wrong_dtype_n_classes, expected_dtype)
+
+
+def test_check_value_ndarray():
+    expected_dtype = np.dtype(np.float64)
+    expected_shape = (5, 1, 2)
+    value_ndarray = np.zeros((5, 1, 2), dtype=expected_dtype)
+
+    allowed_dtypes = [expected_dtype, expected_dtype.newbyteorder()]
+
+    for dt in allowed_dtypes:
+        check_value_ndarray(
+            value_ndarray, expected_dtype=dt, expected_shape=expected_shape
+        )
+
+    with pytest.raises(ValueError, match="Wrong shape.+value array"):
+        check_value_ndarray(
+            value_ndarray, expected_dtype=expected_dtype, expected_shape=(1, 2)
+        )
+
+    for problematic_arr in [value_ndarray[:, :, :1], np.asfortranarray(value_ndarray)]:
+        with pytest.raises(ValueError, match="value array.+C-contiguous"):
+            check_value_ndarray(
+                problematic_arr,
+                expected_dtype=expected_dtype,
+                expected_shape=problematic_arr.shape,
+            )
+
+    with pytest.raises(ValueError, match="value array.+incompatible dtype"):
+        check_value_ndarray(
+            value_ndarray.astype(np.float32),
+            expected_dtype=expected_dtype,
+            expected_shape=expected_shape,
+        )
+
+
+def test_check_node_ndarray():
+    expected_dtype = NODE_DTYPE
+
+    node_ndarray = np.zeros((5,), dtype=expected_dtype)
+
+    valid_node_ndarrays = [
+        node_ndarray,
+        get_different_bitness_node_ndarray(node_ndarray),
+    ]
+    valid_node_ndarrays += [
+        arr.astype(arr.dtype.newbyteorder()) for arr in valid_node_ndarrays
+    ]
+    # TODO add test array with padding?
```

Comment:
I don't think it's necessary. The pickle tests should be enough to detect those kinds of problems.

Reply:
I ended up adding more tests before seeing your review ...